receiver/snmpreceiver/scraper.go

```diff
@@ -0,0 +1,559 @@
+	errMsgBadIndexedAttributes          = `problem retrieving SNMP indexed attribute data: %w`
+	errMsgBadIndexedResourceAttributes  = `problem retrieving SNMP indexed resource attribute data: %w`
+	errMsgIndexedAttributesBadValueType = `returned attribute SNMP data type for OID: %s from column OID: %s is not supported`
+)
+
+// snmpScraper handles scraping of SNMP metrics
+type snmpScraper struct {
+	client   client
+	logger   *zap.Logger
+	cfg      *Config
+	settings component.ReceiverCreateSettings
+}
+
+// newScraper creates an initialized snmpScraper
+func newScraper(logger *zap.Logger, cfg *Config, settings component.ReceiverCreateSettings) *snmpScraper {
+	return &snmpScraper{
+		logger:   logger,
+		cfg:      cfg,
+		settings: settings,
+	}
+}
+
+// start gets the client ready
+func (s *snmpScraper) start(ctx context.Context, host component.Host) (err error) {
+	s.client, err = newClient(s.cfg, host, s.settings.TelemetrySettings, s.logger)
+	if err != nil {
+		return err
+	}
+	err = s.client.Connect()
+
+	return
+}
+
+// shutdown closes the client
+func (s *snmpScraper) shutdown(ctx context.Context, host component.Host) (err error) {
+	err = s.client.Close()
+
+	return
+}
+
+// scrape collects and creates OTEL metrics from a SNMP environment
+func (s *snmpScraper) scrape(_ context.Context) (pmetric.Metrics, error) {
+	// Get a basic ResourceMetrics prepped for metrics with no resource attributes
+	resourceMetrics := pmetric.NewResourceMetrics()
+	scopeMetrics := resourceMetrics.ScopeMetrics().AppendEmpty()
+	scopeMetrics.Scope().SetName("otelcol/snmpreceiver")
+	scopeMetrics.Scope().SetVersion(s.settings.BuildInfo.Version)
+	metricSlice := scopeMetrics.Metrics()
+	now := pcommon.NewTimestampFromTime(time.Now())
+
+	metricsProcessed := false
+
+	resourceMetricsMap := map[string]*pmetric.ResourceMetrics{}
+	metricsMap := map[string]*pmetric.Metric{}
+	// Try to scrape scalar OID based metrics.
+	// metricsMap is passed in as a place for any created metrics to live
+	if err := s.scrapeScalarMetrics(now, metricsMap); err != nil {
+		s.logger.Warn("Failed to collect scalar OID metrics", zap.Error(err))
+	} else if len(metricsMap) != 0 {
+		metricsProcessed = true
+		// Add new scalar metrics to the basic ResourceMetrics with no resource attributes
+		for _, value := range metricsMap {
+			value.MoveTo(metricSlice.AppendEmpty())
+		}
+		// Load the map with the basic ResourceMetrics (that possibly
+		// has scalar OID based metrics already on it)
+		resourceMetricsMap[""] = &resourceMetrics
+	}
+
+	// Load this map with the scalar metric map that was just created for the
+	// basic ResourceMetrics. The info in this map is ultimately already
+	// contained within the resourceMetricsMap, but it is more easily accessable
+	// to pull out a specific existing Metric by resource and metric name here
+	metricsByResourceMap := map[string]map[string]*pmetric.Metric{
```

Comment:
This is really difficult for me to parse when I see this type signature.

To me, it looks like the double map here, is just because it's a composite key.

Could this type actually be better represented by a `map[struct]*pmetric.Metric`?

For instance:
https://go.dev/play/p/2lf6Fym1UJZ

Reply:
If for some reason the struct doesn't work it might be worth creating a type alias for readability. 

```go
type metricMap map[string]*pmetric.Metric
...
metricsByResourceMap := map[string]metricMap
```

Reply:
agree with both of you. the double map is confusing to read. i'll change

Reply:
fixed by adding a special helper struct